brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java

```diff
@@ -0,0 +1,200 @@
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.SdkHttpResponse;
+
+/**
+ * Traces AWS Java SDK V2 calls. Adds on the standard zipkin/brave http tags, as well as tags that
+ * align with the XRay data model.
+ *
+ * This implementation creates 2 types of spans to allow for better error visibility.
+ *
+ * The outer span, "Application Span", wraps the whole SDK operation. This span uses the AWS service
+ * as it's name and will NOT have a remoteService configuration, making it a local span. If the
+ * entire operation results in an error then this span will have an error tag with the cause.
+ *
+ * The inner span, "Client Span", is created for each outgoing HTTP request. This span will be of
+ * type CLIENT. The remoteService will be the name of the AWS service, and the span name will be the
+ * name of the operation being done. If the request results in an error then the span will be
+ * tagged with the error. The AWS request ID is added when available.
+ */
+public class TracingExecutionInterceptor implements ExecutionInterceptor {
+  static final ExecutionAttribute<TraceContext> DEFERRED_ROOT_CONTEXT =
+      new ExecutionAttribute<>("DEFERRED_ROOT_CONTEXT");
+  static final ExecutionAttribute<Span> APPLICATION_SPAN =
+      new ExecutionAttribute<>("APPLICATION_SPAN");
+  static final ExecutionAttribute<Span> CLIENT_SPAN =
+      new ExecutionAttribute<>(Span.class.getCanonicalName());
+
+  static final ExecutionAttribute<String> SERVICE_NAME = SdkExecutionAttribute.SERVICE_NAME;
+
+  static final HttpClientAdapter<SdkHttpRequest.Builder, SdkHttpResponse> ADAPTER =
+      new HttpAdapter();
+  static final Propagation.Setter<SdkHttpRequest.Builder, String> SETTER =
+      SdkHttpRequest.Builder::appendHeader;
+
+  final Tracer tracer;
+  final HttpTracing httpTracing;
+  final HttpClientHandler<SdkHttpRequest.Builder, SdkHttpResponse> handler;
+  final TraceContext.Injector<SdkHttpRequest.Builder> injector;
+
+  TracingExecutionInterceptor(HttpTracing httpTracing) {
+    this.httpTracing = httpTracing;
+    this.tracer = httpTracing.tracing().tracer();
+    this.injector = httpTracing.tracing().propagation().injector(SETTER);
+    this.handler = HttpClientHandler.create(httpTracing, ADAPTER);
+  }
+
+  /**
+   * Before the SDK request leaves the calling thread
+   */
+  @Override public void beforeExecution(
+      Context.BeforeExecution context,
+      ExecutionAttributes executionAttributes
+  ) {
+    Span maybeDeferredRootSpan = tracer.nextSpan();
+    if (maybeDeferredRootSpan.context().parentIdAsLong() == 0) { // Deferred to sampling when we have http context
+      // When we rebuild the context in the 2nd put, we lose the reference here which is the one put
+      // into the PendingSpans list, so we need to save it to keep the span finishable
+      executionAttributes.putAttribute(DEFERRED_ROOT_CONTEXT, maybeDeferredRootSpan.context());
```

Comment:
This is one of the changes that probably needs to get backported. Without an explicit save of the context inside the span we lose the weak reference

Reply:
ps wow this is corner case! I think it will be the same for any not yet sampled right? not just when parent is missing..

Reply:
never mind last comment. nextSpan is in use so I see what you are aiming for. that said another way is to propagate the current parent instead right? suppose the impact is that logs will have parent ID until started but then again maybe that is best.

I *thought * that is what was done in apache hc.. propagate the parent until we are ready to sample.